pennylane/transforms/optimization/pattern_matching.py

```diff
@@ -0,0 +1,1638 @@
+                            condition_block = False
+                            break
+
+                    if ([self.node_id_p, self.node_id_c] in matches_scenario_nomatch) and (
+                        condition_block or not match_backward
+                    ):
+                        new_matching_scenario = MatchingScenarios(
+                            circuit_matched_nomatch,
+                            circuit_blocked_nomatch,
+                            pattern_matched_nomatch,
+                            pattern_blocked_nomatch,
+                            new_matches_scenario_nomatch,
+                            counter_scenario + 1,
+                        )
+                        self.matching_list.append_scenario(new_matching_scenario)
+
+        length = max(len(m.match) for m in match_store_list)
+
+        # Store the matches with maximal length.
+        for scenario in match_store_list:
+            if (len(scenario.match) == length) and not any(
+                scenario.match == x.match for x in self.match_final
+            ):
+                self.match_final.append(scenario)
+
+
+def _gate_indices(circuit_matched, circuit_blocked):
+    """Function which returns the list of gates that are not matched and not blocked for the first scenario.
+    Returns:
+        list(int): list of gate id.
+    """
+    gate_indices = []
+
+    for i, (matched, blocked) in enumerate(zip(circuit_matched, circuit_blocked)):
+        if (not matched) and (not blocked):
+            gate_indices.append(i)
+    gate_indices.reverse()
+    return gate_indices
+
+
+class MaximalMatches:  # pylint: disable=too-few-public-methods
+    """
+    Class MaximalMatches allows to sort and store the maximal matches from the list
+    of matches obtained with the pattern matching algorithm.
+    """
+
+    def __init__(self, pattern_matches):
+        """Initialize MaximalMatches with the necessary arguments.
+        Args:
+            pattern_matches (list): list of matches obtained from running the algorithm.
+        """
+        self.pattern_matches = pattern_matches
+
+        self.max_match_list = []
+
+    def run_maximal_matches(self):
+        """Method that extracts and stores maximal matches in decreasing length order."""
+
+        self.max_match_list = [
+            Match(
+                sorted(self.pattern_matches[0].match),
+                self.pattern_matches[0].qubit,
+            )
+        ]
+
+        for matches in self.pattern_matches[1::]:
+            present = False
+            for max_match in self.max_match_list:
+                for elem in matches.match:
+                    if elem in max_match.match and len(matches.match) <= len(max_match.match):
+                        present = True
+            if not present:
+                self.max_match_list.append(Match(sorted(matches.match), matches.qubit))
+
+
+class SubstitutionConfig:  # pylint: disable=too-many-arguments, too-few-public-methods
```

Comment:
Might be a good candidate for a dataclass.

Reply:
💯